### ISSUE TITLE
Upgrade to Typescript 4

### ISSUE THREAD

**Issue by ActuallyACat** [cultureamp/kaizen-design-system#1769]
- Need to communicate a timeline

**Comment by dmisdm** [comment]
Did some initial research on this:

Upgrading to TS 4.3.5 on murmur and perf-ui gave us these compilation errors:

### Murmur

2 errors found within external library frontend-build, so we first need a PR on that to fix it.
```
node_modules/@cultureamp/frontend-build/lib/build.ts:23:7 - error TS2794: Expected 1 arguments, but got 0. Did you forget to include 'void' in your type argument to 'Promise'?

23       resolve()
         ~~~~~~~~~

  node_modules/typescript/lib/lib.es2015.promise.d.ts:33:34
    33     new <T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~
    An argument for 'value' was not provided.

node_modules/@cultureamp/frontend-build/lib/dev.ts:73:5 - error TS2794: Expected 1 arguments, but got 0. Did you forget to include 'void' in your type argument to 'Promise'?

73     resolve()
       ~~~~~~~~~

  node_modules/typescript/lib/lib.es2015.promise.d.ts:33:34
    33     new <T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~
    An argument for 'value' was not provided.


Found 2 errors.
```


### Perf-ui

15 errors found only within src (nothing external), all easily fixable.

```
src/api/__tests__/performanceCycles.test.ts:21:12 - error TS2790: The operand of a 'delete' operator must be optional.

21     delete reviews[0].user?.job_title_name
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/api/admin/__tests__/performanceCycles.test.ts:25:12 - error TS2790: The operand of a 'delete' operator must be optional.

25     delete reviews[0].manager?.job_title_name
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/api/admin/__tests__/performanceCycles.test.ts:26:12 - error TS2790: The operand of a 'delete' operator must be optional.

26     delete reviews[0].waiting_on_manager?.job_title_name
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/components/ButtonDeprecated/__tests__/ButtonDeprecated.test.tsx:14:47 - error TS2794: Expected 1 arguments, but got 0. Did you forget to include 'void' in your type argument to 'Promise'?

14     new Promise((resolve) => setTimeout(() => resolve(), 1000))
                                                 ~~~~~~~~~

  node_modules/typescript/lib/lib.es2015.promise.d.ts:33:34
    33     new <T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~
    An argument for 'value' was not provided.

src/components/FilterContainer/FilterContainer.tsx:56:40 - error TS2345: Argument of type '{ day: string; month: string; year: string; }' is not assignable to parameter of type 'DateTimeFormatProps'.
  Types of property 'year' are incompatible.
    Type 'string' is not assignable to type '"numeric" | "2-digit" | undefined'.

56       startDate: formatDate(startDate, DATE_FORMAT),
                                          ~~~~~~~~~~~

src/components/FilterContainer/FilterContainer.tsx:57:36 - error TS2345: Argument of type '{ day: string; month: string; year: string; }' is not assignable to parameter of type 'DateTimeFormatProps'.

57       endDate: formatDate(endDate, DATE_FORMAT),
                                      ~~~~~~~~~~~

src/components/FilterContainer/FilterContainer.tsx:209:35 - error TS2345: Argument of type '(DropdownOption<T> | undefined)[]' is not assignable to parameter of type 'DropdownOption<T>[] | (() => DropdownOption<T>[])'.
  Type '(DropdownOption<T> | undefined)[]' is not assignable to type 'DropdownOption<T>[]'.
    Type 'DropdownOption<T> | undefined' is not assignable to type 'DropdownOption<T>'.
      Type 'undefined' is not assignable to type 'DropdownOption<T>'.

209     useState<DropdownOption<T>[]>(mappedVals)
                                      ~~~~~~~~~~

src/components/MarkdownText/MarkdownText.tsx:72:10 - error TS2548: Type 'ReactNode' is not an array type or does not have a '[Symbol.iterator]()' method that returns an iterator.

72       ...children
            ~~~~~~~~

src/containers/TaskList/TaskListContext.tsx:43:32 - error TS2548: Type '(string | number)[] | undefined' is not an array type or does not have a '[Symbol.iterator]()' method that returns an iterator.

43         dismissedRequests: [...state.dismissedRequests, action.payload],
                                  ~~~~~~~~~~~~~~~~~~~~~~~

src/domainHooks/navigation/useNavigationData.ts:349:26 - error TS2774: This condition will always return true since this function is always defined. Did you mean to call it instead?

349               condition: isAdmin && allow_admin_feedback_delete,
                             ~~~~~~~

src/entry/OneOnOneSpacePage/OneOnOneSpacePage.tsx:102:35 - error TS2548: Type '({ __typename?: "OneOnOne" | undefined; } & OneOnOnePage_OneOnOneFragment)[] | null | undefined' is not an array type or does not have a '[Symbol.iterator]()' method that returns an iterator.

102   const oneOnOnes = [oneOnOne, ...historyOneOnOnes]
                                      ~~~~~~~~~~~~~~~~

src/entry/TeamFeedbackRequest/TeamFeedbackRequest.tsx:54:6 - error TS2548: Type 'TeamMember[] | undefined' is not an array type or does not have a '[Symbol.iterator]()' method that returns an iterator.

54   ...team?.details?.teamMembers,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~

src/entry/TeamFeedbackRequest/TeamFeedbackRequest.tsx:55:6 - error TS2548: Type 'TeamMember[] | undefined' is not an array type or does not have a '[Symbol.iterator]()' method that returns an iterator.

55   ...team?.details?.teamLeads,
        ~~~~~~~~~~~~~~~~~~~~~~~~

src/utils/analytics/treasureData.ts:75:5 - error TS2794: Expected 1 arguments, but got 0. Did you forget to include 'void' in your type argument to 'Promise'?

75     resolve()
       ~~~~~~~~~

  node_modules/typescript/lib/lib.es2015.promise.d.ts:33:34
    33     new <T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T>;
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~
    An argument for 'value' was not provided.

src/utils/reflux.ts:76:27 - error TS2548: Type 'any[] | undefined' is not an array type or does not have a '[Symbol.iterator]()' method that returns an iterator.

76   console.log("args:", ...args)
                             ~~~~


Found 15 errors.
```

**Comment by jasononeil** [comment]
Everyone is now on Typescript 4, except us! Let's reopen this.

Will allow us to also fix #1834 and maybe others
